iroh-relay/src/protos/relay.rs

```diff
@@ -536,161 +279,198 @@ impl Frame {
     }
 }
 
-/// Receives the next frame and matches the frame type. If the correct type is found returns the content,
-/// otherwise an error.
-#[cfg(any(test, feature = "server"))]
-pub(crate) async fn recv_frame<E, S: Stream<Item = Result<Frame, E>> + Unpin>(
-    frame_type: FrameType,
-    mut stream: S,
-) -> Result<Frame, E>
-where
-    RecvError: Into<E>,
-{
-    match stream.next().await {
-        Some(Ok(frame)) => {
-            if frame_type != frame.typ() {
-                return Err(UnexpectedFrameSnafu {
-                    got: frame.typ(),
-                    expected: frame_type,
-                }
-                .build()
-                .into());
+impl ClientToRelayMsg {
+    pub(crate) fn typ(&self) -> FrameType {
+        match self {
+            Self::SendPacket { .. } => FrameType::SendPacket,
+            Self::Ping { .. } => FrameType::Ping,
+            Self::Pong { .. } => FrameType::Pong,
+        }
+    }
+
+    pub(crate) fn to_bytes(&self) -> BytesMut {
+        self.write_to(BytesMut::with_capacity(self.encoded_len()))
+    }
+
+    /// Encodes this frame for sending over websockets.
+    ///
+    /// Specifically meant for being put into a binary websocket message frame.
+    pub(crate) fn write_to<O: BufMut>(&self, mut dst: O) -> O {
+        dst = self.typ().write_to(dst);
+        match self {
+            Self::SendPacket { dst_key, packet } => {
+                dst.put(dst_key.as_ref());
+                dst.put(packet.as_ref());
+            }
+            Self::Ping(data) => {
+                dst.put(&data[..]);
+            }
+            Self::Pong(data) => {
+                dst.put(&data[..]);
             }
-            Ok(frame)
         }
-        Some(Err(err)) => Err(err),
-        None => Err(RecvError::from(std::io::Error::new(
-            std::io::ErrorKind::UnexpectedEof,
-            "expected frame".to_string(),
-        ))
-        .into()),
+        dst
+    }
+
+    pub(crate) fn encoded_len(&self) -> usize {
+        let payload_len = match self {
+            Self::Ping(_) | Self::Pong(_) => 8,
+            Self::SendPacket { packet, .. } => {
+                32 // node id
+                + packet.len()
+            }
+        };
+        1 // frame type (all frame types currently encode as 1 byte varint)
```

Comment:
Can we `debug_assert!` this somehow?  I guess it's hard to know what the largest FrameType is going to be?

Reply:
Let me improve code cohesion by adding `FrameType::encoded_len`, so it's right next to the frame type encoding code.